pyomo/core/base/indexed_component_slice.py

```diff
@@ -399,13 +421,20 @@ def __init__(self, component_slice, advance_iter=_advance_iter,
         call_stack = self._slice._call_stack
         call_stack_len = self._slice._len
         self._iter_stack = [None]*call_stack_len
+        # Initialize the base of the `_iter_stack`:
         if call_stack[0][0] == IndexedComponent_slice.slice_info:
+            # Base is a generator for the "lowest-level slice"
             self._iter_stack[0] = _slice_generator(
                 *call_stack[0][1], iter_over_index=self._iter_over_index)
+            # call_stack[0][1] is a (fixed, sliced, ellipsis) tuple, where
+            # fixed and sliced are dicts.
         elif call_stack[0][0] == IndexedComponent_slice.set_item:
             assert call_stack_len == 1
             # defer creating the iterator until later
             self._iter_stack[0] = _NotIterable # Something not None
+            # Initializing to _NotIterable is necessary to avoid an
+            # error when we walk to the base of the call stack, but why
+            # would `set_item` appear at the base of the call stack?
```

Comment:
Still don't really understand why iterator creation is deferred. Why would `set_item` appear on the base of the stack.